python/cudf/cudf/core/scalar.py

```diff
@@ -0,0 +1,242 @@
+
+    def __sub__(self, other):
+        return self._scalar_binop(other, "__sub__")
+
+    def __rsub__(self, other):
+        return self._scalar_binop(other, "__rsub__")
+
+    def __mul__(self, other):
+        return self._scalar_binop(other, "__mul__")
+
+    def __rmul__(self, other):
+        return self._scalar_binop(other, "__rmul__")
+
+    def __truediv__(self, other):
+        return self._scalar_binop(other, "__truediv__")
+
+    def __floordiv__(self, other):
+        return self._scalar_binop(other, "__floordiv__")
+
+    def __rtruediv__(self, other):
+        return self._scalar_binop(other, "__rtruediv__")
+
+    def __mod__(self, other):
+        return self._scalar_binop(other, "__mod__")
+
+    def __divmod__(self, other):
+        return self._scalar_binop(other, "__divmod__")
+
+    def __and__(self, other):
+        return self._scalar_binop(other, "__and__")
+
+    def __xor__(self, other):
+        return self._scalar_binop(other, "__or__")
+
+    def __pow__(self, other):
+        return self._scalar_binop(other, "__pow__")
+
+    def __gt__(self, other):
+        return self._scalar_binop(other, "__gt__")
+
+    def __lt__(self, other):
+        return self._scalar_binop(other, "__lt__")
+
+    def __ge__(self, other):
+        return self._scalar_binop(other, "__ge__")
+
+    def __le__(self, other):
+        return self._scalar_binop(other, "__le__")
+
+    def __eq__(self, other):
+        return self._scalar_binop(other, "__eq__")
+
+    def __ne__(self, other):
+        return self._scalar_binop(other, "__ne__")
+
+    def __round__(self, n):
+        return self._scalar_binop(n, "__round__")
+
+    # Scalar Unary Operations
+    def __abs__(self):
+        return self._scalar_unaop("__abs__")
+
+    def __ceil__(self):
+        return self._scalar_unaop("__ceil__")
+
+    def __floor__(self):
+        return self._scalar_unaop("__floor__")
+
+    def __invert__(self):
+        return self._scalar_unaop("__invert__")
+
+    def __neg__(self):
+        return self._scalar_unaop("__neg__")
+
+    def __repr__(self):
+        return f"Scalar({self.value}, dtype={self.dtype})"
+
+    def _binop_result_dtype_or_error(self, other, op):
+        # import pdb
+        # pdb.set_trace()
+        if op in ["__eq__", "__ne__", "__lt__", "__gt__", "__le__", "__ge__"]:
+            return np.bool
+
+        out_dtype = get_allowed_combinations_for_operator(
+            self.dtype, other.dtype, op
+        )
+
+        # datetime handling
+        if out_dtype in "Mm":
+            if self.dtype.char in "Mm" and other.dtype.char not in "Mm":
+                return self.dtype
+            elif other.dtype.char in "Mm" and self.dtype.char not in "Mm":
+                return other.dtype
+            else:
+                if op == '__sub__' and self.dtype.char == other.dtype.char == 'M':
+                    res, _ = np.datetime_data(max(self.dtype, other.dtype))
+                    return np.dtype('m8'+f"[{res}]")
```

Comment:
I'm a bit confused here, it looks like we get the max unit of the datetime type, then construct a new datetime dtype from that max unit. Isn't that the same as using `max(self.dtype, other.dtype)`?

Reply:
This is the special case when subtracting a datetime (`M`) from another datetime (`M`)  and yielding a timedelta. In that case we want a timedelta dtype (`m`) with the resolution that is the max of the resolutions of the operand datetimes.

Reply:
Got it, thanks!